# Специалист по Data Science

## НИУ ВШЭ, 2024-25 учебный год

## Групповой проект. Exploratory Data Analysis
Выполнили:

Демидов Константин

Быкова Анастасия

## Общая информация

__Дата выдачи:__ 16.01.2025

__Защита проекта:__ не раннее 03.02.2025

## Формат выполнения

Групповой проект представляет собой творческое инициативное групповое задание, которое выполняется студентами в командах от 2 до 4 человек. В то же время, следует помнить, что чем больше в команде людей, тем выше предъявляются требования в целом к команде, а также к выполняему ею проекту.

При оценивании результатов выполнения проекта командой, в обязательном порядке проверяется и учитывается в оценке вклад ***каждого участника*** в общую работу команды. Вследствие этого, крайне важно ответственно подойти к выбору сокомандников, распределению обязанностей внутри команды, а также пресечению наличия "мертвых душ" в команде.

Запись составов команд осуществляется по ссылке в таблице:

https://docs.google.com/spreadsheets/d/1M7aEmCB5GZYf4cF0PptVJCGhjup6jeALlIcsefxqM60/edit?usp=sharing


## Оценивание и штрафы



Групповой проект оценивается исходя из устной защиты (презентации) командой выполненного ими проекта перед комиссией. Комиссия составляется из преподавателей курса, ассистентов курса, а также приглашенных экспертов из индустрии.

В рамках каждой защиты комиссия выставляет два типа оценок.

Первая оценка — это оценка для всего проекта в целом. Она ставится комиссией, исходя из глубины проработки командой предметной области, сложности применяемых в проекте инструментов, а также прочих аспектов, связанных с содержанием выполняемого в проекте ТЗ. При выставлении этой оценки комиссия ориентируется на критерии, более подробно представленные ниже в разделе "Критерии оценки проекта".

Вторая оценка — это индивидуальная оценка каждого участника команды. Именно эта оценка выставляется впоследствии в ведомость. Индивидуальная оценка участника команды определяется комиссией, исходя из того, насколько участник ориентируется в выполненном командой проекте, как он отвечает на адресованные персонально ему вопросы комиссии, насколько в принципе прослеживается его роль в проекте, а также исходя из других аспектов, связанных с участием конкретного человека в деятельности команды. Индивидуальная оценка не может превосходить оценку команды за проект в целом (первую оценку), однако, в случае когда участник спокойно и корректно отвечает на все адресованные ему вопросы, свободно ориентируется в материале и содержании выполненного проекта, а также когда его роль в работе команды не вызывает сомнений, — в таком случае, индивидуальная оценка участника становится равной оценке для всего проекта (т.е. в идеальном случае — вторая оценка равна первой оценке).

Во избежание субъективности и/или предвзятости оценки тех или иных членов комиссии в отношении тех или иных команд, а также каких-либо иных проявлений неравных условий — распределение команд между комиссиями осуществляется абсолютно рандомно с использованием средств библиотеки Numpy :)) — в качестве рэндом сида выступает дата, в которую производится распределение, например, `np.random.seed(16012025)`.

Обратите, пожалуйста, внимание на то, что бОльшая часть комиссий осуществляет приём проектов в очной форме, однако могут присутствовать и комиссии, которые будут работать дистанционно с использованием средств видеосвязи. В случае попадания команды путем случайного распределения в комиссию, работающую в очной форме, *не допускается* какое-либо изменение формата проведения для команды на онлайн-режим, равно как и любые другие перемещения команд между комиссиями (смены комиссий), для сохранения честности и непредвзятости случайного распределения.

Явка студента на защиту проекта является обязательной; вклад студента без его присутствия не оценивается и автоматически является равным нулю. Для дистанционных комиссий, дополнительным требованием является обязательное наличие включенной камеры в течение всего времени нахождения в конференции

## Описание задания

<img src="https://pp.userapi.com/c854528/v854528797/c62ce/FiPEibYUxJc.jpg" width="40%">

Netflix за последние 5-10 лет обзавелись большим количеством зрителей. С увеличением числа зрителей увеличилось и разнообразие шоу. Соответственно, перед аналитиками из киноиндустрии встала задача исследования данных с рейтингами различных сериалов.  

В данном задании вам предстоит провести визуальный анализ датасета **1000 Netflix Shows** (по состоянию на 11.06.2017) и сделать выводы.  



### Описание признаков


* `title` - название шоу.
* `rating` - рейтинг шоу. Например: G, PG, TV-14, TV-MA.
* `ratingLevel` - описание рейтинговой группы и особенностей шоу.
* `ratingDescription` - рейтинг шоу, закодированный числом.
* `release year` - год выпуска шоу.
* `user rating score` - оценка пользователей.
* `user rating size` - общий рейтинг пользователей.

### Требования к проекту



В качестве результата выполнения группового проекта ваша команда должна получить презентацию и защитить ее перед комиссией.

Оформление презентации остаётся полностью на ваше усмотрение, но помните, что результат должен быть релевантен для демонстрации бизнес-заказчику — комиссию, принимающую вашу работу, правильнее всего воспринимать именно в таком качестве. Например, вставлять в презентацию строчки кода или злоупотреблять скринами блокнота не рекомендуется.

С точки зрения концепции выполнения проекта — вам необходимо принять на себя роль аналитиков: провести работу над признаками, исследовать информацию содержащуюся в датасете,  выявить тенденции, тренды, факты из данных, а также, конечно, презентовать всё это в понятном виде.

Фактически, можно воспринимать этот проект в следующем ключе: к вам пришел некий бизнес-заказчик — например, это может быть непосредственно представитель Netflix, которые хотят улучшить какие-то процессы; или какая-то компания, которая хочет снять какой-то новый проект и/или продать какой-то свой уже готовый продукт Netflix'у; или это может быть некая компания конкурент Netflix'a, которая заинтересована в общем исследовании рынка; или абсолютно любой другой стейхолдер в рамках данной отрасли — и вот этот заказчик просит вас проанализировать данные и извлечь на основе них какие-то полезные и значимые бизнес-инсайты для него. Разумеется, чем глубже, чем осмысленнее и чем нетривиальнее будут эти выводы, тем больше они вам заплатят :)


### Примеры вопросов и тем


Чтобы вы не начинали работу полностью с чистого листа, мы подобрали для вас некоторое количество опорных вопросов, с которых вы можете начать ваш первичный анализ.

**Внимание:** Эти вопросы лишь *приблизительные* и дают только *начальный вектор* для вашего исследования. Хотя качественно проработанные ответы на них и могут принести вам значительное количество первичных инсайтов и полезной информации — не стоит строить весь ваш проект вокруг только ответов на них. По возможности постарайтесь усовершенствовать и развить темы, затрагивающиеся в данных вопросах, либо вообще найти другие слои, которые в рамках них не рассматриваются (на самом деле, таковых довольно много!)


Итак, сами вопросы:

- Сколько рейтинговых групп представлено в данных?
- Как рейтинговые группы отличаются друг от друга?
- Какие оценки пользователи ставят чаще всего?
- Какие оценки пользователи ставят в среднем?
- В какие годы были запущены шоу?
- Можно ли сделать вывод, что 2017 год успешнее для Netflix, чем 2016?
- Выберите любое шоу (например, самое понравившееся) и составьте его описательный портрет. На какие вопросы о нем вам было бы интересно ответить, если бы вы были заказчиком? Ответьте на эти вопросы :)
- Какие существуют признаки, не представленные в данных, но существенно влияющие на имеющиеся признаки?
- Как формируются оценки?
- Каким образом собирались сами данные?



Повторимся еще раз: это лишь *примерные вопросы*, для того чтобы дать вам начальный вектор, в котором вы могли бы двигаться. Вы можете отвечать или не отвечать на них по своему усмотрению, а также действовать таким образом и в том направлении, как полностью посчитаете нужным!



### Полезные замечания и советы

- Следите за тем, чтобы все графики на слайдах можно было прочитать даже без ваших комментариев;
- Не нужно делать просто текстовые слайды;
- Стройте свой анализ, исходя из того, что ваша цель представить полезную информацию для бизнеса;
- Любой хороший Data Scientist должен уметь работать в команде, поэтому распределение ваших усилий в проекте остается за вами;
- Не обязательно ограничиваться представленным датасетом. У вас есть весь интернет, чтобы найти дополнительную информацию и проверить ваши гипотезы. Дерзайте!


## Содержательная часть EDA

### Исходные данные

In [ ]:
import pandas as pd
import requests
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.colors as colors
import numpy as np
from scipy.stats import gaussian_kde

In [ ]:
data = pd.read_excel("NetflixShows.xlsx")
del data['ratingDescription'], data['user rating size']

big_data = pd.read_csv("netflix_titles.csv")

Ссылки на внешние источники датасетов:

https://www.kaggle.com/datasets/shivamb/netflix-shows

### Предобработка датасета

##### Дубликаты
- Почему они возникли?
- Много ли их? В каких группах их больше всего?

На всякий случай приведём тайтлы к одинаковому формату:

In [ ]:
data['title'] = data['title'].apply(lambda x: x.strip().lower() if isinstance(x, str) else x)
big_data['title'] = big_data['title'].apply(lambda x: x.strip().lower() if isinstance(x, str) else x)

In [ ]:
data[data.duplicated()]

title rating  \
50          lucky number slevin      R   
52               grey's anatomy  TV-14   
53                 prison break  TV-14   
56        how i met your mother  TV-PG   
57                 supernatural  TV-14   
..                          ...    ...   
991          dawn of the croods  TV-Y7   
992  alpha and omega: dino digs   TV-G   
995                     the bfg     PG   
996     the secret life of pets     PG   
999             growing up wild      G   

                                           ratingLevel  release year  \
50   strong violence, sexual content and adult lang...          2006   
52   Parents strongly cautioned. May be unsuitable ...          2016   
53   Parents strongly cautioned. May be unsuitable ...          2008   
56   Parental guidance suggested. May not be suitab...          2014   
57   Parents strongly cautioned. May be unsuitable ...          2016   
..                                                 ...           ...   
991             Suitable for children ages 7 and older          2017   
992                             Suitable for all ages.          2016   
995  for action/peril, some scary moments and brief...          2016   
996                     for action and some rude humor          2016   
999                                                NaN          2016   

     user rating score  
50                 NaN  
52                98.0  
53                98.0  
56                94.0  
57                95.0  
..                 ...  
991               72.0  
992                NaN  
995               97.0  
996                NaN  
999               80.0  

[500 rows x 5 columns]

Ровно половина датасета - полные дубликаты.
Причины наличия дубликатов могут быть разные:
- Повторый сбор данных
- Ошибка в процессе сбора данных

Дубли стоит искать только в столбце title, где названия фильмов не могут повторяться.

Все остальные атрибуты являются либо категориальными данными, либо могут иметь неуникальное значение.
Также нашей гипотезой является, что ровно половина дублей означает сбор данных дважды. Проверим это, посчитав количество дублей по фильмам:

In [ ]:
data.groupby('title').size().reset_index().sort_values(by=0, ascending=False)

title  0
4                                   13 reasons why  8
152                                       girlboss  7
317                                   prison break  7
275                                       new girl  6
452                            the vampire diaries  6
..                                             ... ..
221      lego jurassic world: the indominus escape  1
219              lego bionicle: the journey to one  1
218  leapfrog: the amazing alphabet amusement park  1
217                                       lazy eye  1
495                      ìò pai, ìò - look at this  1

[496 rows x 2 columns]

Теория с двойным сбором данных не подтверждается т.к. количество дублей разнится и не равняется двум в большинстве случаев.

Очистим датасет от полных дубликатов:

In [ ]:
data = data.drop_duplicates()

In [ ]:
data.groupby('title').size().reset_index().sort_values(by=0, ascending=False)

title  0
356                                  skins  2
68                              bordertown  2
363              star wars: the clone wars  2
157                             goosebumps  2
337  salvador martinha - tip of the tongue  1
..                                     ... ..
162                              greenleaf  1
161                                 grease  1
160                      grace and frankie  1
159                                 gotham  1
495              ìò pai, ìò - look at this  1

[496 rows x 2 columns]

Однако не все дублирующиеся фильмы отсеились - остались неполные дубли. Посмотрим, что с ними не так:

In [ ]:
data[data['title'].duplicated(keep=False)].sort_values(by='title')

title rating  \
167                 bordertown  TV-14   
449                 bordertown  TV-MA   
568                 goosebumps  TV-Y7   
632                 goosebumps     PG   
151                      skins  TV-MA   
181                      skins  TV-MA   
504  star wars: the clone wars     PG   
512  star wars: the clone wars  TV-PG   

                                           ratingLevel  release year  \
167  Parents strongly cautioned. May be unsuitable ...          2016   
449  For mature audiences.  May not be suitable for...          2016   
568             Suitable for children ages 7 and older          1998   
632  scary and intense creature action and images, ...          2015   
151  For mature audiences.  May not be suitable for...          2013   
181                                                NaN          2017   
504  sci-fi action violence throughout, brief langu...          2008   
512  Parental guidance suggested. May not be suitab...          2014   

     user rating score  
167               86.0  
449                NaN  
568               88.0  
632               90.0  
151                NaN  
181                NaN  
504               57.0  
512               93.0

Что мы видим?

Bordertown - сериал, где серии могут иметь разный возрастной ценз.
Goosebumps - два разных проекта, 1998 года - это старый сериал, а 2015 - фильм.
Skins 2013 года - это сериал, а 2017 - фильм.
Star Wars: The Clone Wars 2008 года - фильм или первый сезон одноименного сериала, 2014 - шестой сезон сериала.

Такие ложные дубли можно объяснить и здесь вопрос тогда об уточнении названия тайтлов.

Также при попытке отсортировать тайтлы натыкаемся на ошибку, что где-то не строковые значения, а float. Найдём все НЕ строки:

In [ ]:
data[data['title'].apply(type) != str]

title rating                                        ratingLevel  \
32   0.03  TV-MA  For mature audiences.  May not be suitable for...   
54  90210  TV-14  Parents strongly cautioned. May be unsuitable ...   

    release year  user rating score  
32          2016                NaN  
54          2013               62.0

Эти значения не являются ошибками тоже.
0.03 это такая запись для бразильского сериала "3%", а 90210 - сериал.
Приведём их в строковый тип данных, а также переименуем 0.03 в 3% для удобства:

In [ ]:
data.loc[:, 'title'] = data['title'].astype(str)
data.loc[data['title'] == '0.03', 'title'] = '3%'

Переименуем столбцы датасета также для удобства:

In [ ]:
data.columns = ['title', 'rating', 'ratingLevel', 'releaseYear', 'userRatingScore']

##### Работа с пропусками

Перед тем как отвечать на следующие вопросы, необходимо проверить датасет на наличие пропусков, которые необходимо устранить.

In [ ]:
data.isna().sum()

title                0
rating               0
ratingLevel         33
releaseYear          0
userRatingScore    244
dtype: int64

Пропуски наблюдаются в описании рейтинга и в оценках пользователей.

Посмотрим, какие у нас есть рейтинговые группы:

In [ ]:
data.groupby('rating')['ratingLevel'].unique().to_frame()

ratingLevel
rating                                                     
G          [General Audiences. Suitable for all ages., nan]
NR                    [This movie has not been rated., nan]
PG        [some thematic elements, rude humor and action...
PG-13     [crude and sexual humor, language and some dru...
R         [strong violence, sexual content and adult lan...
TV-14     [Parents strongly cautioned. May be unsuitable...
TV-G                               [Suitable for all ages.]
TV-MA     [For mature audiences.  May not be suitable fo...
TV-PG     [Parental guidance suggested. May not be suita...
TV-Y                               [Suitable for all ages.]
TV-Y7              [Suitable for children ages 7 and older]
TV-Y7-FV  [Suitable for children ages 7 and older.  Cont...
UR        [This movie has not been rated. Intended for a...

Рейтинг UR можно поменять на R т.к. фильм по описанию предназначен для взрослых, просто не поставлен ценз

In [ ]:
data.loc[:, 'rating'] = data['rating'].replace('UR', 'R')

In [ ]:
data.loc[:, 'ratingLevel'] = data.groupby('rating')['ratingLevel'].transform(
    lambda x: x.fillna(x.mode()[0])
)

Пропуски по возрастным цензам заполнили. Остались оценки пользователей.
Почистим для начала названия тайтлов.

In [ ]:
data[data["title"].str.contains('[^A-Za-z0-9\s\.\:\'\(\)\,\&\?\!\-]') == True]

title    rating  \
21                           ìò pai, ìò - look at this     TV-MA   
31                               operaì¤ìµes especiais     TV-MA   
32                                                  3%     TV-MA   
45                                     o amor no divì£     TV-MA   
145                                             merlì?     TV-PG   
296  gabriel lglesias: i‰ûªm sorry for what i said ...     TV-14   
572                            pokì©mon: indigo league      TV-Y   
594                                       pokì©mon: xy  TV-Y7-FV   

                                           ratingLevel  releaseYear  \
21   For mature audiences.  May not be suitable for...         2007   
31   For mature audiences.  May not be suitable for...         2015   
32   For mature audiences.  May not be suitable for...         2016   
45   For mature audiences.  May not be suitable for...         2016   
145  Parental guidance suggested. May not be suitab...         2015   
296  Parents strongly cautioned. May be unsuitable ...         2016   
572                             Suitable for all ages.         2000   
594  Suitable for children ages 7 and older.  Conte...         2015   

     userRatingScore  
21               NaN  
31               NaN  
32               NaN  
45               NaN  
145              NaN  
296             84.0  
572             74.0  
594             56.0

In [ ]:
manual_corrections_by_index = {
    21: 'io pai, io - look at this',
    31: 'operações especiais',
    45: 'o amor no divã',
    145: 'merlí',
    296: "gabriel iglesias: i'm sorry for what i said",
    572: 'pokémon: indigo league',
    594: 'pokémon: xy'
}


for idx, new_title in manual_corrections_by_index.items():
    data.at[idx, 'title'] = new_title

In [ ]:
data[data["title"].str.contains('[^A-Za-z0-9\s\.\:\'\(\)\,\&\?\!\-]') == True]

title    rating  \
31      operações especiais     TV-MA   
32                       3%     TV-MA   
45           o amor no divã     TV-MA   
145                   merlí     TV-PG   
572  pokémon: indigo league      TV-Y   
594             pokémon: xy  TV-Y7-FV   

                                           ratingLevel  releaseYear  \
31   For mature audiences.  May not be suitable for...         2015   
32   For mature audiences.  May not be suitable for...         2016   
45   For mature audiences.  May not be suitable for...         2016   
145  Parental guidance suggested. May not be suitab...         2015   
572                             Suitable for all ages.         2000   
594  Suitable for children ages 7 and older.  Conte...         2015   

     userRatingScore  
31               NaN  
32               NaN  
45               NaN  
145              NaN  
572             74.0  
594             56.0

##### Получение данных из API

Около 50 % данных по оценкам шоу отсутствует, кроме того существующие оценки даны на 2017 год и в настоящее время могут быть неактуальны.
Ввиду того, что получить недостающие данные непосредственно из Netflix не представляется возможным, в качестве альтернативы соберем данные IMDb.

Одним из доступных источников является, например, известный сайт "Кинопоиск", на котором также представлены и другие полезные характеристики шоу, такие как жанр и страна производства.

Сделаем запрос по API через вспомогательный сервис:

In [ ]:
# from urllib.parse import quote_plus

In [ ]:
# API_KEY = "XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX"

In [ ]:
# all_results = {}
# for idx, row in data.iterrows():
#   title = (row['title'])
#   print(f'Кодируем тайтл "{title}"')
#   title = title.strip().lower()  # Приводим к единому формату

#   encoded_title = quote_plus(title)
#   print(f'Закодированный тайтл для url: {encoded_title}')

#   url = f"https://api.kinopoisk.dev/v1.4/movie/search?page=1&limit=1&query={encoded_title}"
#   print(f'Сформированный url: {url}')

#   print('Делаем запрос...')
#   response = requests.get(url, headers={"X-API-KEY": API_KEY})

#   print('Записываем полученный ответ в json словарь')
#   all_results[title] = response.json()
#   print('------------------------')
#   break

In [ ]:
# df_results = pd.DataFrame.from_dict(all_results, orient='index').reset_index()
# df_results.rename(columns={'index': 'title'}, inplace=True)

In [ ]:
# df_columns = ['title', 'genres', 'countries', 'type', 'description',
#               'rating_imdb', 'votes_imdb', 'movieLength', 'ratingMpaa']

# movies_data = []

# for title, data in all_results.items():
#     if 'docs' in data and len(data['docs']) > 0:
#         doc = data['docs'][0]


#         movie_data = {
#             'title': title,
#             'genres': ', '.join([g['name'] for g in doc.get('genres', [])]),
#             'countries': ', '.join([c['name'] for c in doc.get('countries', [])]),
#             'type': doc.get('type', ''),
#             'description': doc.get('description', ''),
#             'rating_imdb': doc.get('rating', {}).get('imdb', None),
#             'votes_imdb': doc.get('votes', {}).get('imdb', None),
#             'movieLength': doc.get('movieLength', None),
#             'ratingMpaa': doc.get('ratingMpaa', '')
#         }

#         movies_data.append(movie_data)

# api_data = pd.DataFrame(movies_data, columns=df_columns)

In [ ]:
# api_data.to_csv('api_data.csv', index=False, encoding='utf-8-sig')

In [ ]:
api_data = pd.read_csv('api_data.csv')

#### Итоговый датасет для анализа данных

##### Обогащение / объединение признаков

Объединим исходный датасет с дополнительно найденными данными:

In [ ]:
final_data = data.merge(api_data, on='title', how='left')

In [ ]:
final_data = final_data.merge(big_data[['title', 'director', 'cast']], on='title', how='left')

In [ ]:
final_data = final_data.drop(columns=['ratingMpaa']) # есть столбец rating

##### Дополнительные преобразования

Преобразуем оценки IMDb к шкале от 1 до 100:

In [ ]:
final_data['rating_imdb'] = final_data['rating_imdb'] * 10

Обобщим рейтинговые группы (идея из анализа рейтингов):

In [ ]:
rating_dict = {'kids': ['TV-Y', 'TV-Y7', 'TV-Y7-FV'],
               'teens': ['TV-14', 'PG-13'],
               'adults only': ['TV-MA', 'R'],
               'all': ['TV-PG', 'TV-G', 'PG', 'G'],
               'unknown': ['NR']}
reverse_rating_dict = {v: k for k, values in rating_dict.items() for v in values}

In [ ]:
final_data['rating_short'] = final_data['rating'].map(reverse_rating_dict)

"Взрываем" датасет для атомизации данных по жанрам и странам:

In [ ]:
final_data_expanded_by_genres = final_data.assign(genres=final_data['genres'].str.split(', ')).explode('genres')

In [ ]:
final_data_expanded_by_genres = final_data_expanded_by_genres.loc[final_data_expanded_by_genres['rating_imdb'] > 0]

In [ ]:
final_data_expanded_by_countries = final_data.assign(countries=final_data['countries'].str.split(', ')).explode('countries')

In [ ]:
final_data_expanded_by_countries = final_data_expanded_by_countries.loc[final_data_expanded_by_countries['rating_imdb'] > 0]

Итак, в финальном датасете final_data представлены следующие столбцы:
- title (Название тайтла)
- rating (Рейтинговая группа)
- ratingLevel (Описание рейтинга)
- releaseYear (Год выхода)
- userRatingScore (Рейтинг тайтла из изначального датасета)
- genres (Жанры)
- countries (Страны - производители)
- type (Тип проекта)
- description (Описание проекта)
- rating_imdb (Рейтинг IMDb)
- votes_imdb (Количество отзывов IMDb)
- movieLength (Хронометраж)
- director (Режиссёр)
- cast (Актёрский состав)

In [ ]:
final_data.head(3)

title rating  \
0         white chicks  PG-13   
1  lucky number slevin      R   
2       grey's anatomy  TV-14   

                                         ratingLevel  releaseYear  \
0  crude and sexual humor, language and some drug...         2004   
1  strong violence, sexual content and adult lang...         2006   
2  Parents strongly cautioned. May be unsuitable ...         2016   

   userRatingScore                    genres  \
0             82.0         комедия, криминал   
1              NaN  триллер, драма, криминал   
2             98.0          драма, мелодрама   

                               countries       type  \
0                                    США      movie   
1  США, Канада, Великобритания, Германия      movie   
2                                    США  tv-series   

                                         description  rating_imdb  votes_imdb  \
0  Проштрафившиеся агенты ФБР - чернокожие братья...         58.0    165480.0   
1  Слевину не везет — дом опечатан, девушка ушла ...         77.0    328413.0   
2  В центре событий — молодая женщина-хирург Мере...         76.0    353710.0   

   movieLength director                                               cast  \
0        105.0      NaN                                                NaN   
1        110.0      NaN                                                NaN   
2          0.0      NaN  Ellen Pompeo, Sandra Oh, Katherine Heigl, Just...   

  rating_short  
0        teens  
1  adults only  
2        teens

In [ ]:
# final_data.to_csv('final_data.csv', index=False)

Удалим 4 проблемных тайтла с пропусками

In [ ]:
final_data = final_data.dropna(subset=['countries'])

А также дубли которые остались после объединения с api данными

In [ ]:
final_data = final_data.drop_duplicates(subset=['title'])

In [ ]:
final_data

title rating  \
0                 white chicks  PG-13   
1          lucky number slevin      R   
2               grey's anatomy  TV-14   
3                 prison break  TV-14   
4        how i met your mother  TV-PG   
..                         ...    ...   
496            russell madness     PG   
497  wiener dog internationals      G   
498                   pup star      G   
499           precious puppies   TV-G   
500                beary tales   TV-G   

                                           ratingLevel  releaseYear  \
0    crude and sexual humor, language and some drug...         2004   
1    strong violence, sexual content and adult lang...         2006   
2    Parents strongly cautioned. May be unsuitable ...         2016   
3    Parents strongly cautioned. May be unsuitable ...         2008   
4    Parental guidance suggested. May not be suitab...         2014   
..                                                 ...          ...   
496                  some rude humor and sports action         2015   
497          General Audiences. Suitable for all ages.         2015   
498          General Audiences. Suitable for all ages.         2016   
499                             Suitable for all ages.         2003   
500                             Suitable for all ages.         2013   

     userRatingScore                                      genres  \
0               82.0                           комедия, криминал   
1                NaN                    триллер, драма, криминал   
2               98.0                            драма, мелодрама   
3               98.0  боевик, триллер, драма, криминал, детектив   
4               94.0                   комедия, мелодрама, драма   
..               ...                                         ...   
496              NaN                                    семейный   
497              NaN                                    семейный   
498              NaN                           комедия, семейный   
499              NaN                                       драма   
500              NaN                              документальный   

                                 countries       type  \
0                                      США      movie   
1    США, Канада, Великобритания, Германия      movie   
2                                      США  tv-series   
3                      США, Великобритания  tv-series   
4                                      США  tv-series   
..                                     ...        ...   
496                                    США      movie   
497                                    США      movie   
498                                 Канада      movie   
499                                    США      movie   
500                               Германия      movie   

                                           description  rating_imdb  \
0    Проштрафившиеся агенты ФБР - чернокожие братья...         58.0   
1    Слевину не везет — дом опечатан, девушка ушла ...         77.0   
2    В центре событий — молодая женщина-хирург Мере...         76.0   
3    Майкл убежден в невиновности брата, осужденног...         83.0   
4    «Как я встретил вашу маму» — комедия про Теда ...         83.0   
..                                                 ...          ...   
496  История о Расселе – маленьком, но добром терье...         43.0   
497  Что делать, если так ждал Чемпионата мира сред...         40.0   
498                                                NaN         44.0   
499  В Гарлеме неграмотной девушке-подростку, страд...         73.0   
500  Эта история произошла в лесах у подножья Карпа...          0.0   

     votes_imdb  movieLength      director  \
0      165480.0        105.0           NaN   
1      328413.0        110.0           NaN   
2      353710.0          0.0           NaN   
3      611933.0          0.0           NaN   
4      750237.0          0.0           NaN   
..          ...          ...           ...   
49

In [ ]:
df_genre_x_score = final_data_expanded_by_genres[['genres', 'rating_imdb', 'votes_imdb']]

df_genre_x_score = df_genre_x_score.groupby('genres').agg(
    rating_imdb=('rating_imdb', 'mean'),
    votes_imdb=('votes_imdb', 'mean'),
    count=('genres', 'count')
).round(2).reset_index()

df_genre_x_score = df_genre_x_score.sort_values(by='rating_imdb', ascending=False)

df_genre_x_score[df_genre_x_score['count'] == 1]

genres  rating_imdb  votes_imdb  count
9      игра         86.0     13934.0      1
23  ток-шоу         64.0      2175.0      1
18  новости         56.0      5967.0      1

In [ ]:
final_data[final_data['genres'].str.contains('игра|ток-шоу|новости')]

title rating  \
103                        chelsea  TV-MA   
106  the great british baking show  TV-PG   
317                bob the builder   TV-Y   

                                           ratingLevel  releaseYear  \
103  For mature audiences.  May not be suitable for...         2017   
106  Parental guidance suggested. May not be suitab...         2016   
317                             Suitable for all ages.         2015   

     userRatingScore                                             genres  \
103              NaN                                   ток-шоу, комедия   
106             94.0                                  игра, реальное ТВ   
317              NaN  мультфильм, новости, фэнтези, комедия, кримина...   

               countries             type  \
103                  США        tv-series   
106       Великобритания        tv-series   
317  Великобритания, США  animated-series   

                                           description  rating_imdb  \
103                                                NaN         64.0   
106  Участниками телепрограммы становятся десять ку...         86.0   
317  Истории о приключениях строителя Боба и его го...         56.0   

     votes_imdb  movieLength         director  \
103      2175.0          0.0              NaN   
106     13934.0          0.0  Andy Devonshire   
317      5967.0          0.0              NaN   

                                                  cast rating_short  
103                                                NaN  adults only  
106  Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...          all  
317                                                NaN         kids

Также можно проверить, в каких строках у нас имеются нулевые значения

In [ ]:
# final_data[final_data.eq(0).any(axis=1)]
final_data.eq(0).sum()

title                0
rating               0
ratingLevel          0
releaseYear          0
userRatingScore      0
genres               0
countries            0
type                 0
description          0
rating_imdb         14
votes_imdb          14
movieLength        243
director             0
cast                 0
rating_short         0
dtype: int64

Заменим нули на пропуски для упрощения работы

In [ ]:
final_data = final_data.replace(0, np.nan)

In [ ]:
final_data.isna().sum()

title                0
rating               0
ratingLevel          0
releaseYear          0
userRatingScore    239
genres               0
countries            0
type                 0
description         67
rating_imdb         14
votes_imdb          14
movieLength        243
director           413
cast               246
rating_short         0
dtype: int64

In [ ]:
final_data = final_data.fillna({
    'rating_imdb': final_data['rating_imdb'].mean(),
    'votes_imdb': final_data['votes_imdb'].mean()
})

Окончательно заполним пропуски

**Описание**
Ну тут просто добавим заглушку

In [ ]:
final_data['description'] = final_data['description'].fillna('Описание отсутствует')

**Хронометраж шоу**

Тут немного сложнее, заполним средним по сериалам и фильмам

In [ ]:
final_data[['title', 'type', 'movieLength']]

title       type  movieLength
0                 white chicks      movie        105.0
1          lucky number slevin      movie        110.0
2               grey's anatomy  tv-series          NaN
3                 prison break  tv-series          NaN
4        how i met your mother  tv-series          NaN
..                         ...        ...          ...
496            russell madness      movie         92.0
497  wiener dog internationals      movie         93.0
498                   pup star      movie         92.0
499           precious puppies      movie        110.0
500                beary tales      movie         52.0

[492 rows x 3 columns]

In [ ]:
final_data

title rating  \
0                 white chicks  PG-13   
1          lucky number slevin      R   
2               grey's anatomy  TV-14   
3                 prison break  TV-14   
4        how i met your mother  TV-PG   
..                         ...    ...   
496            russell madness     PG   
497  wiener dog internationals      G   
498                   pup star      G   
499           precious puppies   TV-G   
500                beary tales   TV-G   

                                           ratingLevel  releaseYear  \
0    crude and sexual humor, language and some drug...         2004   
1    strong violence, sexual content and adult lang...         2006   
2    Parents strongly cautioned. May be unsuitable ...         2016   
3    Parents strongly cautioned. May be unsuitable ...         2008   
4    Parental guidance suggested. May not be suitab...         2014   
..                                                 ...          ...   
496                  some rude humor and sports action         2015   
497          General Audiences. Suitable for all ages.         2015   
498          General Audiences. Suitable for all ages.         2016   
499                             Suitable for all ages.         2003   
500                             Suitable for all ages.         2013   

     userRatingScore                                      genres  \
0               82.0                           комедия, криминал   
1                NaN                    триллер, драма, криминал   
2               98.0                            драма, мелодрама   
3               98.0  боевик, триллер, драма, криминал, детектив   
4               94.0                   комедия, мелодрама, драма   
..               ...                                         ...   
496              NaN                                    семейный   
497              NaN                                    семейный   
498              NaN                           комедия, семейный   
499              NaN                                       драма   
500              NaN                              документальный   

                                 countries       type  \
0                                      США      movie   
1    США, Канада, Великобритания, Германия      movie   
2                                      США  tv-series   
3                      США, Великобритания  tv-series   
4                                      США  tv-series   
..                                     ...        ...   
496                                    США      movie   
497                                    США      movie   
498                                 Канада      movie   
499                                    США      movie   
500                               Германия      movie   

                                           description  rating_imdb  \
0    Проштрафившиеся агенты ФБР - чернокожие братья...    58.000000   
1    Слевину не везет — дом опечатан, девушка ушла ...    77.000000   
2    В центре событий — молодая женщина-хирург Мере...    76.000000   
3    Майкл убежден в невиновности брата, осужденног...    83.000000   
4    «Как я встретил вашу маму» — комедия про Теда ...    83.000000   
..                                                 ...          ...   
496  История о Расселе – маленьком, но добром терье...    43.000000   
497  Что делать, если так ждал Чемпионата мира сред...    40.000000   
498                               Описание отсутствует    44.000000   
499  В Гарлеме неграмотной девушке-подростку, страд...    73.000000   
500  Эта история произошла в лесах у подножья Карпа...    67.399582   

        votes_imdb  movieLength      director  \
0    165480.000000        105.0           NaN   
1    328413.000000        110.0           NaN   
2    353710.000000          NaN           NaN   
3    611933.000000          NaN           NaN   
4    750237.000000          NaN           NaN   
..             ...          ..

In [ ]:
final_data

title rating  \
0                 white chicks  PG-13   
1          lucky number slevin      R   
2               grey's anatomy  TV-14   
3                 prison break  TV-14   
4        how i met your mother  TV-PG   
..                         ...    ...   
496            russell madness     PG   
497  wiener dog internationals      G   
498                   pup star      G   
499           precious puppies   TV-G   
500                beary tales   TV-G   

                                           ratingLevel  releaseYear  \
0    crude and sexual humor, language and some drug...         2004   
1    strong violence, sexual content and adult lang...         2006   
2    Parents strongly cautioned. May be unsuitable ...         2016   
3    Parents strongly cautioned. May be unsuitable ...         2008   
4    Parental guidance suggested. May not be suitab...         2014   
..                                                 ...          ...   
496                  some rude humor and sports action         2015   
497          General Audiences. Suitable for all ages.         2015   
498          General Audiences. Suitable for all ages.         2016   
499                             Suitable for all ages.         2003   
500                             Suitable for all ages.         2013   

     userRatingScore                                      genres  \
0               82.0                           комедия, криминал   
1                NaN                    триллер, драма, криминал   
2               98.0                            драма, мелодрама   
3               98.0  боевик, триллер, драма, криминал, детектив   
4               94.0                   комедия, мелодрама, драма   
..               ...                                         ...   
496              NaN                                    семейный   
497              NaN                                    семейный   
498              NaN                           комедия, семейный   
499              NaN                                       драма   
500              NaN                              документальный   

                                 countries       type  \
0                                      США      movie   
1    США, Канада, Великобритания, Германия      movie   
2                                      США  tv-series   
3                      США, Великобритания  tv-series   
4                                      США  tv-series   
..                                     ...        ...   
496                                    США      movie   
497                                    США      movie   
498                                 Канада      movie   
499                                    США      movie   
500                               Германия      movie   

                                           description  rating_imdb  \
0    Проштрафившиеся агенты ФБР - чернокожие братья...    58.000000   
1    Слевину не везет — дом опечатан, девушка ушла ...    77.000000   
2    В центре событий — молодая женщина-хирург Мере...    76.000000   
3    Майкл убежден в невиновности брата, осужденног...    83.000000   
4    «Как я встретил вашу маму» — комедия про Теда ...    83.000000   
..                                                 ...          ...   
496  История о Расселе – маленьком, но добром терье...    43.000000   
497  Что делать, если так ждал Чемпионата мира сред...    40.000000   
498                               Описание отсутствует    44.000000   
499  В Гарлеме неграмотной девушке-подростку, страд...    73.000000   
500  Эта история произошла в лесах у подножья Карпа...    67.399582   

        votes_imdb  movieLength      director  \
0    165480.000000        105.0           NaN   
1    328413.000000        110.0           NaN   
2    353710.000000          NaN           NaN   
3    611933.000000          NaN           NaN   
4    750237.000000          NaN           NaN   
..             ...          ..

### Аналитика данных

#### Количество рейтинговых групп

In [ ]:
final_data['rating'].nunique()

12

Всего в датасете представлено 12 рейтинговых групп:

*   7 групп соответствуют системе TVPG и относятся к сериалам и телешоу (TV-Y, TV-Y7, TV-Y7-FV, TV-G, TV-PG, TV-14, TV-MA),
*   4 группы соответствуют системе MPAA и относятся к фильмам (G, PG, PG-13, R),
*   1 группа, обозначающая отсутствие присвоенного рейтинга (NR).

В основе данных систем лежит определение, насколько подходит тот или иной фильм или телепередача для детей определенного возраста. Как правило, универсальных возрастных ограничений для фильмов нет и в каждой стране приняты свои системы.

В США приняты TVPG и MPAA.

Посмотрим, сколько фильмов представлено в каждой рейтинговой группе:

In [ ]:
rating_counts = final_data['rating'].value_counts().to_frame().reset_index()
rating_counts

rating  count
0      TV-14    104
1      TV-MA     79
2         PG     75
3          G     53
4       TV-Y     35
5      TV-PG     32
6       TV-G     29
7   TV-Y7-FV     25
8      TV-Y7     23
9          R     15
10     PG-13     12
11        NR     10

In [ ]:
fig = px.bar(rating_counts, x = 'rating', y='count', title="Распределение количества запущенных шоу по рейтинговым группам")
fig.update_layout(xaxis_title = 'Категория', yaxis_title = 'Количество', font={'size': 16})
fig.show()

In [ ]:
labels = rating_counts['rating']
values = rating_counts['count']
fig = go.Figure()
pull = [0]*len(values)
pull[values.tolist().index(values.max())] = 0.2
init_colors = colors.sequential.Plasma
n_colors = len(labels)
discrete_colors = [init_colors[int(i)] for i in np.linspace(0, len(init_colors) - 1, n_colors)]
# colors.sequential.Turbo
# colors.qualitative.Bold

fig.add_trace(go.Pie(labels = labels, values=values, pull = pull, hole=0.4,
                     marker=dict(colors=discrete_colors, line=dict(color='black', width=0.2))))

# fig.update_layout(font={'size': 14}, template='plotly_dark',  plot_bgcolor='rgba(0, 0, 0, 0)' )
fig.show()

In [ ]:
rating_dict = {'kids': ['TV-Y', 'TV-Y7', 'TV-Y7-FV'],
               'teens': ['TV-14', 'PG-13'],
               'adults only': ['TV-MA', 'R'],
               'all': ['TV-PG', 'TV-G', 'PG', 'G'],
               'unknown': ['NR']}
reverse_rating_dict = {v: k for k, values in rating_dict.items() for v in values}
reverse_rating_dict

{'TV-Y': 'kids',
 'TV-Y7': 'kids',
 'TV-Y7-FV': 'kids',
 'TV-14': 'teens',
 'PG-13': 'teens',
 'TV-MA': 'adults only',
 'R': 'adults only',
 'TV-PG': 'all',
 'TV-G': 'all',
 'PG': 'all',
 'G': 'all',
 'NR': 'unknown'}

In [ ]:
final_data['rating_short'] = final_data['rating'].map(reverse_rating_dict)

In [ ]:
rating_short_counts = final_data['rating_short'].value_counts().to_frame().reset_index()
values = rating_short_counts['count']
labels = rating_short_counts['rating_short']
fig = go.Figure()

init_colors = colors.sequential.Plasma
n_colors = len(labels)
discrete_colors = [init_colors[int(i)] for i in np.linspace(0, len(init_colors) - 1, n_colors)]

fig.add_trace(go.Pie(labels = labels, values = values, hole=0.4,
                     textinfo='label+percent',
                     marker=dict(colors=discrete_colors, line=dict(color='black', width=0.2))))
# fig.update_layout(font={'size': 16}, template='plotly_dark',  plot_bgcolor='rgba(0, 0, 0, 0)', paper_bgcolor='rgba(0, 0, 0, 0)' )
fig.show()

Сделаем некоторое обобщение рейтинговых групп, объединив их в 5:


*   kids - для детей
*   teens - для зрителей с 14 лет
*   adults only - для зрителей с 17 лет
*   all - для зрителей всех возрастов или с одобрения взрослых
*   unknown - рейтинг не определен (оставим его, как есть)

В целом, можно отметить, что контент предназначен для широкой аудитории.

#### Какие оценки пользователи ставят чаще?

- Постройте гистограмму оценок.
- Почему именно такие оценки ставят чаще всего?

Данные, полученные из Кинопоиска, актуальны на январь 2025 г, в то время как данные исходного датасета приведены по состоянию на июнь 2017 г. В связи с этим, мы можем посмотреть изменились ли рейтинги фильмов и сериалов и насколько:

In [ ]:
final_data['score_delta'] = final_data['userRatingScore'] - final_data['rating_imdb']
final_data['score_delta'].describe().to_frame().T

count       mean        std   min  25%  50%   75%   max
score_delta  253.0  10.062456  14.392251 -33.0  1.0  9.0  21.0  40.0

In [ ]:
x_data = final_data['title'].loc[final_data['userRatingScore'] > 0]
y_data = final_data['userRatingScore'].loc[final_data['userRatingScore'] > 0]
y2_data = final_data['rating_imdb'].loc[final_data['userRatingScore'] > 0]

fig = go.Figure()
fig.add_trace(go.Scatter(x=x_data, y=y_data, mode='markers', name = 'Netflix'))
fig.add_trace(go.Scatter(x=x_data, y=y2_data, mode='markers', line=dict(color='green'), name = 'IMDb'))
fig.update_layout(title = 'Сопоставление оценок Netflix и IMDb', xaxis_title = 'Шоу', yaxis_title = 'Средняя оценка',
                  # width = 1000, height = 400, template='plotly_dark',
                  # plot_bgcolor='rgba(0, 0, 0, 0)',
                  legend=dict(
        orientation="h",
        yanchor="bottom",
        y=1.02,
        xanchor="center",
        x=0.9
    ))
fig.update_xaxes(showticklabels=False)
fig.show()

В среднем оценки отличаются на 15%. Оценки IMDb в целом менее положительные, чем оценки Netflix.

In [ ]:
fig = px.histogram(final_data['rating_imdb'],
                   nbins=100,
                   opacity=0.8,
                   histnorm='density',
                   title = 'Распределение значений оценок IMDb')

mean_value = final_data['rating_imdb'].mean()

fig.add_trace(go.Scatter(
    x=[mean_value, mean_value],
    y=[0, 35],
    mode='lines',
    line=dict(color='white', dash='dash'),
    name='mean'
))

fig.update_layout(xaxis_title = 'Значение оценки', yaxis_title = 'Количество',
                  # width = 1000, height = 400, template='plotly_dark',
                  # plot_bgcolor='rgba(0, 0, 0, 0)',
                  legend=dict(
        title_text='',
        orientation="h",
        yanchor="bottom",
        y=1.02,
        xanchor="center",
        x=0.9
    ))

fig.add_annotation(
    x=mean_value, y=25,
    text=f'Mean: {mean_value:.1f}',
    font=dict(
        color='black'
    ),
    showarrow=True,
    arrowhead=1,
    ax=50,
    ay=0,
    # bgcolor='rgba(255, 255, 255, 0.8)',
    bordercolor='white',
    borderwidth=1
)

fig.show()

Чаще всего пользователи ставят хорошие оценки, выше 6.

In [ ]:
low_rated = final_data[final_data["rating_imdb"] < 50]
mid_rated = final_data[(final_data["rating_imdb"] >= 50) & (final_data["rating_imdb"] < 80)]
high_rated = final_data[final_data["rating_imdb"] >= 80]

rating_counts = pd.DataFrame({
    "Категория": ["Низкооцененные", "Среднеоцененные", "Высокооцененные"],
    "Количество": [len(low_rated), len(mid_rated), len(high_rated)]
})

fig = px.pie(
    rating_counts,
    names="Категория",
    values="Количество",
    title="Процентное соотношение рейтингов фильмов",
    hole=0.4
)

fig.show()

#### Основная информация об оценках пользователей
- Чему равны среднее, стандартное отклонение, минимум, максимум, медиана?
- Отличаются ли медиана и среднее?
- Могут ли данные характеристики значительно отличаться? Почему?

In [ ]:
final_data['rating_imdb'].describe().to_frame().T.round(1)

count  mean   std   min   25%   50%   75%   max
rating_imdb  492.0  67.4  12.0  28.0  60.0  67.4  77.0  95.0

Среднее значение может сильно отличаться от медианы, т.к. среднее подвержено влиянию единичных сильно больших или малых значений. В нашем случае таких выборосов в оценках нет, поэтому среднее и медиана относительно близки.

In [ ]:
fig = px.scatter(x=np.log1p(final_data['votes_imdb']), y=final_data['rating_imdb'],trendline='ols', trendline_color_override='red', title="Зависимость средней оценки от количества оценок")
fig.update_layout(xaxis_title = 'Количество оценок', yaxis_title = 'Оценка', font={'size': 12},
                  width = 600, height = 400,
                  # template='plotly_dark',
                  # plot_bgcolor='rgba(0, 0, 0, 0)'
                  )
fig.show()

Помимо распределения оценок мы также можем посмотреть, есть ли зависимость между средней оценкой и количеством голосов. Одназначно можно сказать лишь, что у проектов с большим количеством голосов (порядка миллиона) в основном высокие оценки. Наличие большого количества голосов здесь скорее свидельствует о популярности шоу, которая подтверждается высокой оценкой.

#### Распределение количества запущенных шоу в зависимости от года.

- Наблюдается ли рост?
- Есть ли выбросы?
- Что из себя представляют выбросы?
- Чем могут быть они вызваны?

In [ ]:
print(f"Год самого раннего шоу: {final_data['releaseYear'].min()} \nГод самого позднего шоу: {final_data['releaseYear'].max()}")

Год самого раннего шоу: 1940 
Год самого позднего шоу: 2017


In [ ]:
show_counts = final_data['releaseYear'].value_counts().to_frame().reset_index()

fig = px.bar(show_counts, x = 'releaseYear', y='count', title="Распределение количества запущенных проектов по годам")
fig.update_layout(xaxis_title = 'Год релиза', yaxis_title = 'Количество проектов', font={'size': 14})
fig.show()

In [ ]:
final_data[final_data['releaseYear'] == 1940]

title rating                                ratingLevel  releaseYear  \
413  fantasia      G  General Audiences. Suitable for all ages.         1940   

     userRatingScore                                         genres countries  \
413             61.0  мультфильм, мюзикл, фэнтези, семейный, музыка       США   

        type                                        description  rating_imdb  \
413  cartoon  Необычайный, поражающий воображение мультконце...         77.0   

     votes_imdb  movieLength director cast rating_short  score_delta  
413    104630.0        125.0      NaN  NaN          all        -16.0

В выборке предствлены фильмы от 1940 до 2017 гг. По распределению проектов по годам можно наблюдать рост релизов до 2016 года и затем снижение в 2017, однако, здесь необходимо помнить, что данные на 2017 год присутствуют только за полгода.

In [ ]:
final_data[['title', 'type', 'releaseYear']].loc[final_data['type'] == 'tv-series']

title       type  releaseYear
2                     grey's anatomy  tv-series         2016
3                       prison break  tv-series         2008
4              how i met your mother  tv-series         2014
5                       supernatural  tv-series         2016
6                       breaking bad  tv-series         2013
..                               ...        ...          ...
357                           jessie  tv-series         2015
359                         lab rats  tv-series         2016
404                           bunk'd  tv-series         2016
435  jack and the cuckoo-clock heart  tv-series         2013
492              h2o: just add water  tv-series         2009

[167 rows x 3 columns]

Также, если мы выведем список только сериалов и проверим несколько тайтлов, можем обратить внимание, что год релиза - это не обязательно год выхода первого сезона.

Например, сериал "Как я встретил вашу маму" вышел в 2005 году. В 2014 году вышел последний, 9 сезон.

Посмотрим распределение по годам только для фильмов и мультфильмов:

In [ ]:
show_counts = final_data['releaseYear'].value_counts().to_frame().reset_index()
trace0 = go.Bar(
    x=show_counts['releaseYear'],
    y=show_counts['count'],
    name = 'Все проекты'
)

movie_counts = final_data['releaseYear'].loc[((final_data['type'] == 'movie') | (final_data['type'] == 'cartoon'))].value_counts().to_frame().reset_index()
trace1 = go.Bar(
    x=movie_counts['releaseYear'],
    y=movie_counts['count'],
    name = 'Фильмы и мультфильмы'
)

our_data = [trace0, trace1]

fig = go.Figure(data = our_data)
fig.update_layout(title="Распределение количества запущенных проектов по годам", xaxis_title = 'Год релиза', yaxis_title = 'Количество проектов', font={'size': 14})
fig.update_xaxes(range=[1990, None])
fig.show()

#### Средняя оценка пользователей в 2016 / 2017.
- Можно ли сделать вывод, что 2017 год успешнее для Netflix? ("Успешнее" значит, что пользователи в среднем ставили более высокие оценки)
- Как еще можно оценить "Успешность"?



In [ ]:
score_by_year = final_data[['releaseYear', 'rating_imdb']].groupby('releaseYear').mean('rating_imdb').reset_index()
shows_by_year = final_data['releaseYear'].value_counts().reset_index()

fig = go.Figure()

fig.add_trace(go.Scatter(
    x=score_by_year['releaseYear'],
    y=score_by_year['rating_imdb'],
    mode='lines+markers',
    name='Средний IMDb рейтинг',
    yaxis='y1'
))


fig.add_trace(go.Bar(
    x=shows_by_year['releaseYear'],
    y=shows_by_year['count'],
    name='Количество шоу',
    marker=dict(color='#009639', opacity=0.8),
    yaxis='y2'
))

fig.update_layout(
    title="Распределение количества запущенных шоу по годам и их средний рейтинг",
    xaxis=dict(title="Год выпуска",
               range=[1990, 2018]),

    yaxis=dict(
        title="Средний IMDb рейтинг",
        side="left",
        showgrid=False,
        range=[0, 75]
    ),

    yaxis2=dict(
        title="Количество проектов",
        overlaying="y",
        side="right",
        showgrid=False,
        tickmode="sync",
        range=[0, 150]
    ),

    legend=dict(
        title_text='',
        orientation="h",
        yanchor="bottom",
        y=1.02,
        xanchor="center",
        x=0.9),

    # width = 1000, height = 400,
    # template='plotly_dark',
    # paper_bgcolor='rgba(0,0,0,0)',
    # plot_bgcolor='rgba(0,0,0,0)'
)

fig.show()

2017 год резко проседает по объёму выпущенных шоу т.к. наши данные ограничиваются 06.2017 и полгода ещё впереди

#### Дополнительные графики

In [ ]:
countries_dict = { 'США': 'USA',
    'Канада': 'Canada',
    'Великобритания': 'United Kingdom',
    'Германия': 'Germany',
    'Япония': 'Japan',
    'Австралия': 'Australia',
    'Бельгия': 'Belgium',
    'Дания': 'Denmark',
    'Франция': 'France',
    'Польша': 'Poland',
    'Колумбия': 'Colombia',
    'Бразилия': 'Brazil',
    'Мексика': 'Mexico',
    'Индия': 'India',
    'Португалия': 'Portugal',
    'Новая Зеландия': 'New Zealand',
    'Испания': 'Spain',
    'Грузия': 'Georgia',
    'Гана': 'Ghana',
    'Ирландия': 'Ireland',
    'Болгария': 'Bulgaria',
    'Турция': 'Turkey',
    'Китай': 'China',
    'Тайвань': 'Taiwan',
    'Корея Южная': 'South Korea',
    'Финляндия': 'Finland',
    'Австрия': 'Austria',
    'Кипр': 'Cyprus',
    'Таиланд': 'Thailand',
    'Словения': 'Slovenia',
    'Сингапур': 'Singapore',
    'Сербия и Черногория': 'Serbia and Montenegro',
    'Гонконг': 'Hong Kong',
    'Филиппины': 'Philippines',
    'Италия': 'Italy',
    'Люксембург': 'Luxembourg',
    'Россия': 'Russia',
    'Гватемала': 'Guatemala',
    'Малайзия': 'Malaysia',
    'Аргентина': 'Argentina'}

In [ ]:
final_data_expanded_by_countries['countries_en'] = final_data_expanded_by_countries['countries'].map(countries_dict)

In [ ]:
for_map = final_data_expanded_by_countries.groupby('countries_en')['rating_imdb'].mean().reset_index()
for_map.describe().T

count       mean       std   min   25%   50%       75%   max
rating_imdb   40.0  64.961453  9.278881  41.0  60.0  63.9  68.10227  90.0

In [ ]:
fig = px.choropleth(for_map,
                    locations='countries_en',
                    locationmode='country names',
                    color='rating_imdb',
                    # animation_frame=final_data_expanded_by_countries['releaseYear'],
                    color_continuous_scale="RdYlGn",
                    range_color = [0, 100])
# fig.update_layout(
#     template='plotly_dark',
#     plot_bgcolor='rgba(0, 0, 0, 0)'
# )
fig.show()

##### Популярность жанров
Переменные:
- Жанр
- Средняя суммарная оценка шоу
- Среднее количество отзывов
- Количество шоу по жанрам

In [ ]:
df_genre_x_score = final_data_expanded_by_genres[['genres', 'rating_imdb', 'votes_imdb']]

df_genre_x_score = df_genre_x_score.groupby('genres').agg(
    rating_imdb=('rating_imdb', 'mean'),
    votes_imdb=('votes_imdb', 'mean'),
    count=('genres', 'count')
).round(2).reset_index()

df_genre_x_score = df_genre_x_score.sort_values(by='rating_imdb', ascending=False)

df_genre_x_score[df_genre_x_score['count'] <= 10]


genres  rating_imdb  votes_imdb  count
9          игра        86.00    13934.00      1
20  реальное ТВ        79.00     6988.00      2
0         аниме        77.83    57949.00      6
10      история        77.50   100672.25      4
1     биография        73.10    77933.80     10
4       военный        69.67    22602.83      6
3       вестерн        69.25    33562.00      4
23      ток-шоу        64.00     2175.00      1
18      новости        56.00     5967.00      1

In [ ]:
fig_10 = go.Figure()

fig_10.add_trace(go.Bar(
    x=df_genre_x_score['genres'],
    y=df_genre_x_score['rating_imdb'],
    name="Средняя оценка",
    yaxis='y1'
))

fig_10.add_trace(go.Bar(
    x=df_genre_x_score['genres'],
    y=df_genre_x_score['votes_imdb'],
    name="Количество отзывов",
    yaxis='y2',
    opacity=0.8,
    marker=dict(color='#009639')
))

fig_10.update_layout(
    title="Популярность жанров (Количество отзывов)",
    xaxis=dict(title="Жанр"),
    width = 1200, height = 400,
    # template='plotly_dark',
    # plot_bgcolor='rgba(0, 0, 0, 0)',
    yaxis=dict(
        title="Средняя оценка",
        side='left'
    ),
    yaxis2=dict(
        title="Количество отзывов",
        overlaying='y',
        side='right'
    ),
    barmode='group',
    showlegend=True,

    legend=dict(
        orientation="h",
        yanchor="bottom",
        y=1.02,
        xanchor="center",
        x=0.9
    )
)



fig_10.show()

In [ ]:
fig_11 = go.Figure()

fig_11.add_trace(go.Bar(
    x=df_genre_x_score['genres'],
    y=df_genre_x_score['rating_imdb'],
    name="Средняя оценка",
    yaxis='y1'
))

fig_11.add_trace(go.Bar(
    x=df_genre_x_score['genres'],
    y=df_genre_x_score['count'],
    name="Количество шоу",
    yaxis='y2',
    opacity=0.8,
    marker=dict(color='#EAAA00')
))

fig_11.update_layout(
    title="Популярность жанров (Количество шоу)",
    xaxis=dict(title="Жанр"),
    width = 1200, height = 400,
    # template='plotly_dark',
    # plot_bgcolor='rgba(0, 0, 0, 0)',
    yaxis=dict(
        title="Средняя оценка",
        side='left'
    ),
    yaxis2=dict(
        title="Количество шоу",
        overlaying='y',
        side='right'
    ),
    barmode='group',
    showlegend=True,

    legend=dict(
        orientation="h",
        yanchor="bottom",
        y=1.02,
        xanchor="center",
        x=0.9
    )
)

fig_11.show()

##### Оценка от длительности фильма

In [ ]:
df_length_x_score = final_data[['title', 'rating_imdb', 'movieLength']]
df_length_x_score = df_length_x_score.dropna()
df_length_x_score

title  rating_imdb  movieLength
0                 white chicks    58.000000        105.0
1          lucky number slevin    77.000000        110.0
12                  death note    45.000000        101.0
15                  the hunter    67.000000        104.0
21          masha and the bear    67.000000         89.0
..                         ...          ...          ...
496            russell madness    43.000000         92.0
497  wiener dog internationals    40.000000         93.0
498                   pup star    44.000000         92.0
499           precious puppies    73.000000        110.0
500                beary tales    67.399582         52.0

[249 rows x 3 columns]

In [ ]:
fig = go.Figure()


fig = go.Figure()

fig.add_trace(go.Scatter(
    x=final_data["movieLength"],
    y=final_data["rating_imdb"],
    mode="markers",
    marker=dict(size=10, color="blue", opacity=0.6),
    name="Фильмы",
    text=final_data["title"],
    hoverinfo="text"
))


fig.update_layout(
    title="Зависимость оценки от длительности фильма",
    xaxis_title="Длительность фильма (мин)",
    yaxis_title="Рейтинг IMDb",
    showlegend=False
)

fig.show()

##### Распределение оценок (Фильмы/Сериалы)

In [ ]:
final_data[['type', 'rating_imdb']]

type  rating_imdb
0        movie    58.000000
1        movie    77.000000
2    tv-series    76.000000
3    tv-series    83.000000
4    tv-series    83.000000
..         ...          ...
496      movie    43.000000
497      movie    40.000000
498      movie    44.000000
499      movie    73.000000
500      movie    67.399582

[492 rows x 2 columns]

In [ ]:
final_data['type'] = final_data['type'].apply(lambda x: 'movie' if 'movie' in x else 'series')

In [ ]:
fig = px.violin(
    final_data,
    y="rating_imdb",
    color="type",
    violinmode='overlay',
    points="all"
)

fig.update_layout(
    violingap=0.5
)

fig.show()

In [ ]:
movies = final_data[final_data['type'] == 'movie']
series = final_data[final_data['type'] == 'series']

fig = go.Figure()

fig.add_trace(go.Violin(
  x=movies['type'],
  y=movies['rating_imdb'],
  name='Фильмы',
  box_visible=True,
  meanline_visible=True,
  points='outliers',
  fillcolor='royalblue',
  line_color='darkblue',
  opacity=0.8
))

# Виолончельный график для сериалов
fig.add_trace(go.Violin(
  x=series['type'],
  y=series['rating_imdb'],
  name='Сериалы',
  box_visible=True,
  meanline_visible=True,
  points='outliers',
  fillcolor='seagreen',
  line_color='darkgreen',
  opacity=0.8
))

# Улучшаем оформление графика
fig.update_layout(
  title='Распределение оценок по типам тайтлов',
  yaxis_title='Рейтинг IMDb',
  xaxis_title='Тип тайтла',
  legend_title="Категории",
  width = 800, height = 800,
  violingap=0,
  # template='plotly_dark',
  # paper_bgcolor='rgba(0, 0, 0, 0)',
  # plot_bgcolor='rgba(0, 0, 0, 0)',
  font={'size': 18}
)

fig.show()


##### Сочетания жанров

In [ ]:
all_genres = final_data['genres'].apply(lambda x: x.split(', '))

unique_genres = set()
for row in all_genres:
  for g in row:
      unique_genres.add(g.strip())

unique_genres = list(unique_genres)
len(unique_genres)

28

In [ ]:
matrix = pd.DataFrame(
  0,
  index=unique_genres,
  columns=unique_genres
)

In [ ]:
from itertools import combinations

In [ ]:
for genre_list in all_genres:
  for g1, g2 in combinations(genre_list, 2):
      g1, g2 = g1.strip(), g2.strip()
      matrix.loc[g1, g2] += 1
      matrix.loc[g2, g1] += 1

In [ ]:
size = 800
fig_15 = go.Figure(data=go.Heatmap(
  x=matrix.columns,
  y=matrix.index,
  z=matrix,
  xgap=1,
  ygap=1,
  colorbar=dict(title="Сочетаний"),
  colorscale=[
        [0, "#16272f"],
        [1, "#048036"]
    ]
    ))

fig_15.update_layout(
  # title="Сочетания жанров в различных тайтлах",
  width=size, height=size,

  template='plotly_dark',
  paper_bgcolor='rgba(0,0,0,0)',
  plot_bgcolor='rgba(0,0,0,0)'
)

fig_15.show()

In [ ]:
matrix.unstack().nlargest(15).reset_index()

level_0      level_1    0
0       комедия     семейный  133
1      семейный      комедия  133
2      семейный   мультфильм  133
3    мультфильм     семейный  133
4      семейный  приключения  122
5   приключения     семейный  122
6    мультфильм  приключения  111
7   приключения   мультфильм  111
8       комедия   мультфильм  103
9    мультфильм      комедия  103
10      комедия  приключения  100
11  приключения      комедия  100
12     семейный      фэнтези   81
13      фэнтези     семейный   81
14   мультфильм      фэнтези   77

In [ ]:
matrix = np.log1p(matrix)
# matrix = pd.DataFrame(np.tril(matrix))
fig_15 = go.Figure(data=go.Heatmap(
  x=matrix.columns,
  y=matrix.index,
  xgap=1,
  ygap=1,
  z=matrix,
  colorbar=dict(title="Сочетаний (log)"),
  colorscale=[
        [0, "#16272f"],
        [1, "#048036"]
    ]
    ))

fig_15.update_layout(
  # title="Сочетания жанров в различных тайтлах",
  width=size, height=size,

  template='plotly_dark',
  paper_bgcolor='rgba(0,0,0,0)',
  plot_bgcolor='rgba(0,0,0,0)'
)

fig_15.show()

#### Топ проектов

Какие проекты наиболее успешны?

In [ ]:
final_data.sort_values(by = 'rating_imdb', ascending=False)

title rating  \
6                        breaking bad  TV-MA   
11                           sherlock  TV-14   
51                  the office (u.s.)  TV-14   
98                   better call saul  TV-MA   
24             hunter x hunter (2011)     NR   
..                                ...    ...   
449            the hero of color city      G   
485                inspector gadget 2      G   
441        alpha and omega: dino digs   TV-G   
115  amy schumer: the leather special  TV-MA   
443                           hatched  TV-Y7   

                                           ratingLevel  releaseYear  \
6    For mature audiences.  May not be suitable for...         2013   
11   Parents strongly cautioned. May be unsuitable ...         2016   
51   Parents strongly cautioned. May be unsuitable ...         2012   
98   For mature audiences.  May not be suitable for...         2016   
24                      This movie has not been rated.         2013   
..                                                 ...          ...   
449          General Audiences. Suitable for all ages.         2014   
485          General Audiences. Suitable for all ages.         2003   
441                             Suitable for all ages.         2016   
115  For mature audiences.  May not be suitable for...         2017   
443             Suitable for children ages 7 and older         2013   

     userRatingScore                                             genres  \
6               97.0                           криминал, драма, триллер   
11              95.0                 детектив, триллер, драма, криминал   
51              97.0                                            комедия   
98              95.0                                    драма, криминал   
24              57.0    аниме, мультфильм, комедия, приключения, боевик   
..               ...                                                ...   
449              NaN         мультфильм, фэнтези, приключения, семейный   
485              NaN    фантастика, боевик, комедия, криминал, семейный   
441              NaN  мультфильм, фэнтези, комедия, приключения, сем...   
115              NaN                                            комедия   
443              NaN             приключения, фантастика, боевик, ужасы   

                     countries    type  \
6                          США  series   
11         Великобритания, США  series   
51                         США  series   
98                         США  series   
24              Япония, Польша  series   
..                         ...     ...   
449    США, Индия, Корея Южная  series   
485             Австралия, США   movie   
441  США, Канада, Индия, Китай  series   
115                        США   movie   
443             Великобритания   movie   

                                           description  rating_imdb  \
6    Школьный учитель химии Уолтер Уайт узнаёт, что...         95.0   
11   События разворачиваются в наши дни. Он прошел ...         91.0   
51   Сериал о трудовых буднях небольшого региональн...         90.0   
98   История об испытаниях и невзгодах, которые при...         90.0   
24   Лицензия члена организации Охотников дает множ...         90.0   
..                                                 ...          ...   
449  Мультик поведает нам историю о разноцветных ме...         39.0   
485  Трепещите, злодеи! Гроза криминала и заступник...         35.0   
441  Миллионы лет тому назад, когда на планете жили...         34.0   
115  Эми Шумер рассказывает о сексе, свиданиях и не...         32.0   
443                               Описание отсутствует         28.0   

     votes_imdb  movieLength     director  \
6     2264211.0          NaN          NaN   
11    1032442.0          NaN          NaN   
51     755754.0          NaN          NaN   
98     703083.0          NaN          NaN   
24     154159.0          NaN          NaN   
..          ...          ...          ...   
449       754.0         77.0

In [ ]:
data.sort_values(by = 'userRatingScore', ascending=False).head(15)

title rating  \
41                      13 reasons why  TV-MA   
350                               lost  TV-14   
10                    once upon a time  TV-PG   
64                             friends  TV-14   
72             orange is the new black  TV-MA   
27                           the flash  TV-PG   
25                  marvel's iron fist  TV-MA   
88                        finding dory     PG   
62                          family guy  TV-MA   
63                      criminal minds  TV-14   
3                         prison break  TV-14   
2                       grey's anatomy  TV-14   
8                     the walking dead  TV-MA   
86                            zootopia     PG   
310  law & order: special victims unit  TV-14   

                                           ratingLevel  releaseYear  \
41   For mature audiences.  May not be suitable for...         2017   
350  Parents strongly cautioned. May be unsuitable ...         2010   
10   Parental guidance suggested. May not be suitab...         2016   
64   Parents strongly cautioned. May be unsuitable ...         2003   
72   For mature audiences.  May not be suitable for...         2016   
27   Parental guidance suggested. May not be suitab...         2016   
25   For mature audiences.  May not be suitable for...         2017   
88                              mild thematic elements         2016   
62   For mature audiences.  May not be suitable for...         2015   
63   Parents strongly cautioned. May be unsuitable ...         2016   
3    Parents strongly cautioned. May be unsuitable ...         2008   
2    Parents strongly cautioned. May be unsuitable ...         2016   
8    For mature audiences.  May not be suitable for...         2015   
86       some thematic elements, rude humor and action         2016   
310  Parents strongly cautioned. May be unsuitable ...         2015   

     userRatingScore  
41              99.0  
350             98.0  
10              98.0  
64              98.0  
72              98.0  
27              98.0  
25              98.0  
88              98.0  
62              98.0  
63              98.0  
3               98.0  
2               98.0  
8               98.0  
86              97.0  
310             97.0

In [ ]:
final_data

title rating  \
0                 white chicks  PG-13   
1          lucky number slevin      R   
2               grey's anatomy  TV-14   
3                 prison break  TV-14   
4        how i met your mother  TV-PG   
..                         ...    ...   
496            russell madness     PG   
497  wiener dog internationals      G   
498                   pup star      G   
499           precious puppies   TV-G   
500                beary tales   TV-G   

                                           ratingLevel  releaseYear  \
0    crude and sexual humor, language and some drug...         2004   
1    strong violence, sexual content and adult lang...         2006   
2    Parents strongly cautioned. May be unsuitable ...         2016   
3    Parents strongly cautioned. May be unsuitable ...         2008   
4    Parental guidance suggested. May not be suitab...         2014   
..                                                 ...          ...   
496                  some rude humor and sports action         2015   
497          General Audiences. Suitable for all ages.         2015   
498          General Audiences. Suitable for all ages.         2016   
499                             Suitable for all ages.         2003   
500                             Suitable for all ages.         2013   

     userRatingScore                                      genres  \
0               82.0                           комедия, криминал   
1                NaN                    триллер, драма, криминал   
2               98.0                            драма, мелодрама   
3               98.0  боевик, триллер, драма, криминал, детектив   
4               94.0                   комедия, мелодрама, драма   
..               ...                                         ...   
496              NaN                                    семейный   
497              NaN                                    семейный   
498              NaN                           комедия, семейный   
499              NaN                                       драма   
500              NaN                              документальный   

                                 countries    type  \
0                                      США   movie   
1    США, Канада, Великобритания, Германия   movie   
2                                      США  series   
3                      США, Великобритания  series   
4                                      США  series   
..                                     ...     ...   
496                                    США   movie   
497                                    США   movie   
498                                 Канада   movie   
499                                    США   movie   
500                               Германия   movie   

                                           description  rating_imdb  \
0    Проштрафившиеся агенты ФБР - чернокожие братья...    58.000000   
1    Слевину не везет — дом опечатан, девушка ушла ...    77.000000   
2    В центре событий — молодая женщина-хирург Мере...    76.000000   
3    Майкл убежден в невиновности брата, осужденног...    83.000000   
4    «Как я встретил вашу маму» — комедия про Теда ...    83.000000   
..                                                 ...          ...   
496  История о Расселе – маленьком, но добром терье...    43.000000   
497  Что делать, если так ждал Чемпионата мира сред...    40.000000   
498                               Описание отсутствует    44.000000   
499  В Гарлеме неграмотной девушке-подростку, страд...    73.000000   
500  Эта история произошла в лесах у подножья Карпа...    67.399582   

        votes_imdb  movieLength      director  \
0    165480.000000        105.0           NaN   
1    328413.000000        110.0           NaN   
2    353710.000000          NaN           NaN   
3    611933.000000          NaN           NaN   
4    750237.000000          NaN           NaN   
..             ...          ...           ...   
496     730.00000

#### Любимое шоу

- The Batman (Сериал)
- Breaking Bad (Фильм)

In [ ]:
batman = final_data[final_data['title'] == 'the batman']
breaking_bad = final_data[final_data['title'] == 'breaking bad']

In [ ]:
batman.merge(breaking_bad, how='outer')

title rating                                        ratingLevel  \
0  breaking bad  TV-MA  For mature audiences.  May not be suitable for...   
1    the batman  TV-Y7             Suitable for children ages 7 and older   

   releaseYear  userRatingScore                             genres countries  \
0         2013             97.0           криминал, драма, триллер       США   
1         2007             77.0  детектив, криминал, драма, боевик       США   

     type                                        description  rating_imdb  \
0  series  Школьный учитель химии Уолтер Уайт узнаёт, что...         95.0   
1   movie  После двух лет поисков правосудия на улицах Го...         78.0   

   votes_imdb  movieLength director  \
0   2264211.0          NaN      NaN   
1    853084.0        176.0      NaN   

                                                cast rating_short  score_delta  
0  Bryan Cranston, Aaron Paul, Anna Gunn, Dean No...  adults only          2.0  
1                                                NaN         kids         -1.0

In [ ]:
batman_year = 2007
bb_year = 2013

batman_genres = {'детектив', 'криминал', 'драма', 'боевик'}
bb_genres = {'криминал', 'драма', 'триллер'}

batman_rating = 'TV-Y7'
bb_rating = 'TV-MA'

both_country = 'США'

batman_type = 'movie'
bb_type = 'series'

batman_imdb = 78.0
breaking_bad = 95.0

batman_votes = 853084.0
bb_votes = 2264211.0

batman_length = 176.0

In [ ]:
score_by_year = final_data[['releaseYear', 'rating_imdb']].groupby('releaseYear').mean('rating_imdb').reset_index()
shows_by_year = final_data['releaseYear'].value_counts().reset_index()

years = score_by_year['releaseYear']
avg_ratings = score_by_year['rating_imdb']
show_counts = shows_by_year['releaseYear']
num_shows = shows_by_year['count']

colors_by_shows = ['purple' if year == batman_year else
          'aqua' if year == bb_year else
          'green' for year in show_counts]

colors_by_score = ['purple' if year == batman_year else
          'aqua' if year == bb_year else
          'green' for year in years]

fig = go.Figure()

fig.add_trace(go.Scatter(
    x=score_by_year['releaseYear'],
    y=score_by_year['rating_imdb'],
    mode='lines+markers',
    name='Средний IMDb рейтинг',
    yaxis='y1',
    marker=dict(color=colors_by_score)
))


fig.add_trace(go.Bar(
    x=shows_by_year['releaseYear'],
    y=shows_by_year['count'],
    name='Количество шоу',
    marker=dict(color=colors_by_shows),
    yaxis='y2'
))

fig.update_layout(
    title="Распределение количества запущенных шоу по годам и их средний рейтинг",
    xaxis=dict(title="Год выпуска",
               range=[1990, 2018]),

    yaxis=dict(
        title="Средний IMDb рейтинг",
        side="left",
        showgrid=False,
        range=[0, 75]
    ),

    yaxis2=dict(
        title="Количество проектов",
        overlaying="y",
        side="right",
        showgrid=False,
        tickmode="sync",
        range=[0, 150]
    ),

    legend=dict(
        title_text='',
        orientation="h",
        yanchor="bottom",
        y=1.02,
        xanchor="center",
        x=0.9),
    width = 1000, height = 400,
    # template='plotly_dark',

    # paper_bgcolor='rgba(0,0,0,0)',
    # plot_bgcolor='rgba(0,0,0,0)'
)

fig.show()

#### Feature engineering

In [ ]:
# ordinal encoding
rating_order = {
    'TV-Y': 1,
    'TV-Y7': 2,
    'TV-Y7-FV': 3,
    'TV-G': 4,
    'G': 5,
    'TV-PG': 6,
    'PG': 7,
    'TV-14': 8,
    'PG-13': 9,
    'R': 10,
    'TV-MA': 11,
    'NR': 6  # NR можно присвоить средний рейтинг или удалить
}

# Применяем Ordinal Encoding
# final_data['rating'] = final_data['rating'].map(rating_order)
# final_data = final_data.drop(columns=['ratingLevel'])
final_data

title rating  \
0                 white chicks  PG-13   
1          lucky number slevin      R   
2               grey's anatomy  TV-14   
3                 prison break  TV-14   
4        how i met your mother  TV-PG   
..                         ...    ...   
496            russell madness     PG   
497  wiener dog internationals      G   
498                   pup star      G   
499           precious puppies   TV-G   
500                beary tales   TV-G   

                                           ratingLevel  releaseYear  \
0    crude and sexual humor, language and some drug...         2004   
1    strong violence, sexual content and adult lang...         2006   
2    Parents strongly cautioned. May be unsuitable ...         2016   
3    Parents strongly cautioned. May be unsuitable ...         2008   
4    Parental guidance suggested. May not be suitab...         2014   
..                                                 ...          ...   
496                  some rude humor and sports action         2015   
497          General Audiences. Suitable for all ages.         2015   
498          General Audiences. Suitable for all ages.         2016   
499                             Suitable for all ages.         2003   
500                             Suitable for all ages.         2013   

     userRatingScore                                      genres  \
0               82.0                           комедия, криминал   
1                NaN                    триллер, драма, криминал   
2               98.0                            драма, мелодрама   
3               98.0  боевик, триллер, драма, криминал, детектив   
4               94.0                   комедия, мелодрама, драма   
..               ...                                         ...   
496              NaN                                    семейный   
497              NaN                                    семейный   
498              NaN                           комедия, семейный   
499              NaN                                       драма   
500              NaN                              документальный   

                                 countries    type  \
0                                      США   movie   
1    США, Канада, Великобритания, Германия   movie   
2                                      США  series   
3                      США, Великобритания  series   
4                                      США  series   
..                                     ...     ...   
496                                    США   movie   
497                                    США   movie   
498                                 Канада   movie   
499                                    США   movie   
500                               Германия   movie   

                                           description  rating_imdb  \
0    Проштрафившиеся агенты ФБР - чернокожие братья...    58.000000   
1    Слевину не везет — дом опечатан, девушка ушла ...    77.000000   
2    В центре событий — молодая женщина-хирург Мере...    76.000000   
3    Майкл убежден в невиновности брата, осужденног...    83.000000   
4    «Как я встретил вашу маму» — комедия про Теда ...    83.000000   
..                                                 ...          ...   
496  История о Расселе – маленьком, но добром терье...    43.000000   
497  Что делать, если так ждал Чемпионата мира сред...    40.000000   
498                               Описание отсутствует    44.000000   
499  В Гарлеме неграмотной девушке-подростку, страд...    73.000000   
500  Эта история произошла в лесах у подножья Карпа...    67.399582   

        votes_imdb  movieLength      director  \
0    165480.000000        105.0           NaN   
1    328413.000000        110.0           NaN   
2    353710.000000          NaN           NaN   
3    611933.000000          NaN           NaN   
4    750237.000000          NaN           NaN   
..             ...          ...           ...   
496     730.00000

## Критерии оценки проекта

Оценка за групповой проект в целом — то есть первая из двух оценок — выставляется по 10 балльной шкале, исходя из 5 критериев. Также комиссия на своё усмотрение может добавить (но не снизить) некоторое количество бонусных баллов за рамками данных критериев.

Каждый из критериев будет перечислен и рассмотрен ниже

### Критерий 1. Информативность техническая – необходимо выявить нетривиальные зависимости и тенденции в данных (до 2 баллов)

Данный критерий оценивает, насколько вы продвинулись технически с точки зрения выявления каких-то паттернов, тенденций и зависимостей в данных. Чем более сложные инструменты вы примените в ходе выполнения проекта и чем более сложные и нетривиальные зависимости получите, тем выше будет оценка у вашей команды по данному критерию!

### Критерий 2. Информативность аналитическая – необходимо интерпретировать данные и делать выводы полезные для бизнеса  (до 2 баллов)

Данный критерий оценивает прикладной смысл проделанных вами действий, интерпретацию технических инструментов, умение подсветить полезность и значимость полученных выводов с точки зрения бизнеса, умение переложить полученные результаты на исходные бизнес-цели и "продать" то, что вы сделали.


### Критерий 3. Визуализация – оценивается техническое исполнение графиков (до 2 баллов)

Данный критерий оценивает сложность и техничность ваших визуализаций. Визуализации должны быть наглядными, красивыми, понятными, легко-интерпертируемыми, нетривиальными с точки зрения технического исполнения и выводов, которые они несут в себе.


*Максимальный балл по данному критерию (2 балла) нельзя получить, если в работе используются лишь базовые библиотеки визуализаций: Matplotlib и Seaborn*

### Критерий 4. Предобработка датасета – очистка/замена данных (до 1 балла)

Данный критерий оценивает полноту всех проделанных действий с точки зрения подготовки данных и первичной предобработки датасета.


### Критерий 5. Обогащение данными – добавление новых признаков/датасетов (до 3 баллов)

Данный критерий оценивает полноту использования средств feature engineering на основе имеющихся данных, а также средств обогащения данных путем дополнения имеющегося датасета значимой информацией из связанных сторонних датасетов, различных внешних ресурсов и других источников. Обратите внимание, что связанная информация извне не должна замещать или подменять собой исходную задачу и исходные данные в представленном датасете, а должна лишь дополнять и обогощать его. Корректность и глубина проработки взаимодействия со внешними ресурсами и оценивается в рамках данного критерия!

*Максимальный балл по данному критерию (**3 балла**) можно получить только при условии корректной и значимой работы с как минимум двумя источниками внешних данных (двумя внешними датасетами), а также при условии наличия как минимум двух примеров валидного и осмысленного применения средств feature engineering*

*По данному критерию можно получить **2 балла** только при условии корректной и значимой работы с как минимум одним источником внешних данных (внешним датасетом), а также при условии наличия как минимум одного примера валидного и осмысленного применения средств feature engineering*